isolate/mtn.go

```diff
@@ -0,0 +1,294 @@
+	defer rh.Body.Close()
+	if doErr != nil {
+		if rh.StatusCode == 400 {
+			errResp := AllocError{}
+			decoder := json.NewDecoder(rh.Body)
+			rErr := decoder.Decode(&errResp)
+			if rErr != nil {
+				return nil, fmt.Errorf(
+					"%s ERROR: Cant allocate. Request error: %s. Internal error: %s. Caused: %s.",
+					time.Now().UTC().Format(time.RFC3339),
+					doErr,
+					errResp.Message[0],
+					errResp.Cause.Message[0],
+				)
+			} else {
+				return nil, fmt.Errorf(
+					"%s ERROR: Cant allocate. Request error: %s. Body parse error: %s.",
+					time.Now().UTC().Format(time.RFC3339),
+					doErr,
+					rErr,
+				)
+			}
+		}
+		return nil, doErr
+	}
+	r := make(map[string][]Allocation)
+	jresp := []RawAlloc{}
+	decoder := json.NewDecoder(rh.Body)
+	rErr := decoder.Decode(&jresp)
+	if rErr != nil {
+		return nil, rErr
+	}
+	for _, a := range jresp {
+		r[a.Network] = append(r[a.Network], Allocation{a.Porto.Net, a.Porto.Hostname, a.Porto.Ip, a.Id, false})
+	}
+	return r, nil
+}
+
+func (c *MtnState) RequestAllocs(netid string) (map[string]Allocation, error) {
+	r := make(map[string]Allocation)
+	ctx, cancel := context.WithTimeout(context.Background(), 20 * time.Second)
+	defer cancel()
+	jsonBody := PostAllocreq{netid, c.Cfg.Ident, c.Cfg.SchedLabel}
+	txtBody, mrshErr := json.Marshal(jsonBody)
+	if mrshErr != nil {
+		return nil, mrshErr
+	}
+	for i := 0; i < c.Cfg.Allocbuffer; i++ {
+		req, nrErr := http.NewRequest("POST", c.Cfg.Url, bytes.NewBuffer(txtBody))
+		if nrErr != nil {
+			return nil, nrErr
+		}
+		for header, value := range c.Cfg.Headers {
+			req.Header.Set(header, value)
+		}
+		req.Header.Set("Content-Type", "application/json")
+		req = req.WithContext(ctx)
+		rh, doErr := http.DefaultClient.Do(req)
+		defer rh.Body.Close()
+		if doErr != nil {
+			return nil, doErr
+		}
+		jresp := RawAlloc{}
+	        decoder := json.NewDecoder(rh.Body)
+		rErr := decoder.Decode(&jresp)
+		if rErr != nil {
+			return nil, rErr
+		}
+		r[jresp.Id] = Allocation{jresp.Porto.Net, jresp.Porto.Hostname, jresp.Porto.Ip, jresp.Id, false}
+	}
+	return r, nil
+
+}
+
+func (c *MtnState) BindAllocs(netId string) error {
+	c.Lock()
+	defer c.Unlock()
+	if _, p := c.Pool[netId]; p {
+		c.Pool[netId].Lock()
+		defer c.Pool[netId].Unlock()
+		if (len(c.Pool[netId].Allocations) - c.Pool[netId].Reserved) > c.Cfg.Allocbuffer {
+			c.Pool[netId].Reserved += c.Cfg.Allocbuffer
+			return nil
+		} else {
+			allocs, reqErr := c.RequestAllocs(netId)
+			if reqErr != nil {
+				return reqErr
+			}
+			c.Pool[netId].Reserved += c.Cfg.Allocbuffer
+			for id, alloc := range allocs {
+				c.Pool[netId].Allocations[id] = alloc
+			}
+			return nil
+		}
+	} else {
+		allocs, reqErr := c.RequestAllocs(netId)
+		if reqErr != nil {
+			return reqErr
+		}
+		newPool := IdState{*new(sync.Mutex), c.Cfg.Allocbuffer, allocs}
+		c.Pool[netId] = &newPool
+		return nil
+	}
+}
+
+func (c *MtnState) UnbindAllocs(netId string) {
+	c.Pool[netId].Lock()
```

Comment:
Is this method called under c.Pool level Lock

Reply:
tnx, i tiny rework locks usage now